alpa/device_mesh.py

```diff
@@ -364,13 +381,56 @@ def init_broadcast_communicator(group_name, comm_key, world_size,
         g._get_nccl_broadcast_communicator(  # pylint: disable=protected-access
             comm_key, world_size, device_ids, devices_global_rank, nccl_uid)
 
+    def xla_nccl_send_tile(self, uuid: int, offset: Sequence[slice],
+                           dst_rank: int, dst_gpu_idx: int, group_name: str):
+
+        if global_config.pipeline_use_signal_send_recv:
+            signal = self.signal_tensors[uuid % len(self.local_devices)]
+            col.send_multigpu(signal,
+                              dst_rank,
+                              dst_gpu_idx,
+                              group_name,
+                              start_pos=0,
+                              n_elements=1)
+            return
+
+        tensor_shape = self.buffers[uuid].shape
+        if is_continuous_subset(offset, tensor_shape):
+            start_pos, n_elements = \
```

Comment:
Why there is no `slice_shape == tensor_shape` branch just like what in `cupy_nccl_send_tile`?

Reply:
It is because our new api takes starting point and number of elements as input arguments. When `slice_shape == tensor_shape`, starting point is zero and the number of elements is exactly the size of the tensor. You could see from my code, as long as it is contiguous, we could use the same code to deal with.

Reply:
Because these two code `xla_nccl_**` and `cupy_nccl_**` are not very similar to each other and we may fully delete cupy later, I did not merge them together.